dair_pll/experiment.py

```diff
@@ -340,6 +342,26 @@ def train_epoch(self, data: DataLoader, system: System,
         avg_loss = cast(Tensor, sum(losses) / len(losses))
         return avg_loss
 
+    def calculate_loss_no_grad_step(self, data: DataLoader, system: System) \
```

Comment:
Seems almost copy-paste from `train_epoch`. Probably better to have 1 function that takes a flag as an argument for whether to take the gradient steps or not -- e.g. by making the `Optimizer` argument to `train_epoch` an `Optional[Optimizer]`, and take gradient steps iff it's not `None`.

Reply:
Agreed.  Done just like your suggestion.